cli/src/main/java/com/liferay/blade/cli/command/WatchCommand.java

```diff
@@ -75,102 +98,307 @@ private void _addError(String prefix, String msg) {
+		throws IOException {
+
+		Files.walkFileTree(
+			basePath,
+			new SimpleFileVisitor<Path>() {
+
+				@Override
+				public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes basicFileAttributes)
+					throws IOException {
+
+					for (PathMatcher pathMatcher : ignorePathMatchers) {
+						if (pathMatcher.matches(path)) {
+							return FileVisitResult.SKIP_SUBTREE;
+						}
+					}
+
+					_registerDirectory(watchService, watchKeys, path);
+
+					return FileVisitResult.CONTINUE;
+				}
+
+			});
+	}
+
+	private void _watch(
+			Path watchPath, Map<String, Path> projectPaths, List<String> fastPaths, List<String> ignorePaths,
+			boolean deploy)
+		throws InterruptedException {
+
 		Thread watchThread = new Thread() {
 
 			@Override
 			public void run() {
 				BladeCLI bladeCLI = getBladeCLI();
 
 				try {
+					FileSystem fileSystem = FileSystems.getDefault();
+
+					final WatchService watchService = fileSystem.newWatchService();
+
+					final Map<WatchKey, Path> watchKeys = new HashMap<>();
+
+					final List<PathMatcher> ignorePathMatchers = _getPathMatchers(
+						watchPath, ignorePaths.toArray(new String[0]));
+
+					final List<PathMatcher> fastPathMatchers = _getPathMatchers(
+						watchPath, fastPaths.toArray(new String[0]));
+
+					_walkAndRegisterDirectories(watchService, watchKeys, watchPath, ignorePathMatchers);
+
 					final GradleExec gradleExec = new GradleExec(bladeCLI);
 
-					Set<String> projectPaths = projectOutputFiles.keySet();
+					if (deploy) {
+						bladeCLI.out("Deploying...  To skip initial deployment, use `blade watch -s`");
 
-					Stream<String> stream = projectPaths.stream();
+						gradleExec.executeTask("deploy", false);
+					}
 
-					String assembleTasks = stream.collect(Collectors.joining(":assemble "));
+					while (true) {
+						WatchKey watchKey;
 
-					String assembleTaskPath = assembleTasks + ":assemble";
+						try {
+							watchKey = watchService.take();
+						}
+						catch (InterruptedException ie) {
+							return;
+						}
 
-					gradleExec.executeTask(assembleTaskPath, watchPath.toFile(), false);
+						Path dir = watchKeys.get(watchKey);
 
-					stream = projectPaths.stream();
+						if (dir == null) {
+							bladeCLI.error("WatchKey not recognized!!");
 
-					String watchTasks = stream.collect(Collectors.joining(":watch "));
+							continue;
+						}
 
-					String watchTaskPath = watchTasks + ":watch --continuous --no-rebuild --stacktrace";
+						for (WatchEvent<?> event : watchKey.pollEvents()) {
+							WatchEvent.Kind<?> kind = event.kind();
 
-					gradleExec.executeTask(watchTaskPath, watchPath.toFile(), false);
-				}
-				catch (Exception e) {
-					String message = e.getMessage();
+							Path path = (Path)event.context();
 
-					if (message == null) {
-						message = "Gradle task failed.";
-					}
+							Path resolvedPath = dir.resolve(path);
+
+							Path projectPath = _getGradleProjectPath(watchPath, resolvedPath, projectPaths);
```

Comment:
Perhaps this should be moved down to line 340, to avoid calculating `projectPath` unnecessarily.

Reply:
Good point.